runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -328,6 +328,22 @@ public void close() throws Exception {
         }
       }
     }
+
+    if (currentOutputWatermark != Long.MAX_VALUE) {
```

Comment:
You're right, `close()` can be called before the watermark reaches the end when a job is cancelled. Fixing.